src/odb.c

```diff
@@ -669,6 +671,97 @@ void git_odb_free(git_odb *db)
 	GIT_REFCOUNT_DEC(db, odb_free);
 }
 
+typedef struct {
+	git_oid oid;
+	git_array_oid_t parents;
+} git_odb__graft;
+
+int git_odb__graft_register(git_odb *odb, const git_oid *oid, git_array_oid_t parents)
+{
+	int res = 0, error = 0;
+	size_t k, i;
+	git_odb__graft *graft;
+	git_oid *parent_oid;
+
+	assert(odb && oid);
+
+	graft = git__calloc(1, sizeof(*graft));
+	GITERR_CHECK_ALLOC(graft);
+
+	git_array_init_to_size(graft->parents, git_array_size(parents));
+	git_array_foreach(parents, i, parent_oid) {
+		git_oid *id = git_array_alloc(graft->parents);
+		GITERR_CHECK_ALLOC(id);
+
+		git_oid_cpy(id, parent_oid);
+	}
+
+	git_oid_cpy(&graft->oid, oid);
+	k = git_oidmap_put(odb->grafts, &graft->oid, &res);
+	if (res < 0) {
+		error = -1;
+		goto cleanup;
+	}
+
+	git_oidmap_set_value_at(odb->grafts, k, graft);
+
+cleanup:
+	return error;
+}
+
+int git_odb__graft_unregister(git_odb *odb, const git_oid *oid)
+{
+	assert(odb && oid);
+
+	git_oidmap_delete(odb->grafts, oid);
+
+	return 0;
+}
+
+static int graft_for_oid(git_odb__graft **out, git_odb *odb, const git_oid *oid)
+{
+	size_t pos;
+
+	assert(out && odb && oid);
+
+	pos = git_oidmap_lookup_index(odb->grafts, oid);
+	if (!git_oidmap_valid_index(odb->grafts, pos)) {
+		return GIT_ENOTFOUND;
+	}
+
+	*out = git_oidmap_value_at(odb->grafts, pos);
+
+	return 0;
+}
+
+int git_odb__graft_object(void *obj, git_odb *odb)
```

Comment:
Why do you use a void pointer here? Shouldn't that be a `git_object *` pointer instead?

Reply:
Was expecting a "layering violation", because of `git_odb_object`. Also, I think I saw a precedent, hence thought it would be fine. If it's not, it should be changed to `git_commit *` because you only graft onto commits.

Reply:
The question that occurs to me then is why this function is called `git_odb__graft_object` when you're only supposed to graft commits ;)